core/src/main/scala/org/apache/spark/sql/CachedDataFrame.scala

```diff
@@ -412,7 +412,6 @@ object CachedDataFrame
     val buffer = new Array[Byte](4 << 10) // 4K
     // final output is written to this buffer
     val output = new ByteBufferDataOutput(4 << 10,
-      DirectBufferAllocator.instance(),
```

Comment:
If a user has not set -memory-size we will be using more memory than -Xmx. JVM heap plus some off-heap memory. Would that be alright?  
A cleaner solution would be to use accounting in HeapBufferAllocator. 
The problem is we don't know which pool to access there. Storage or execution.